cloudpickle/cloudpickle.py

```diff
@@ -628,12 +628,16 @@ def save_global(self, obj, name=None, pack=struct.pack):
         The name of this method is somewhat misleading: all types get
         dispatched here.
         """
+        if obj.__module__ == "__main__":
+            return self.save_dynamic_class(obj)
```

Comment:
Does this mean `save_dynamic_class` is also used for *functions*? If so, could you rename that method and update its docstring?

Reply:
I don't think functions get dispatched here.  The dispatch entries for `save_global` are:

```python
     dispatch[type] = save_global
     dispatch[types.ClassType] = save_global
```

whereas function-likes are dispatched via:

```python
dispatch[types.FunctionType] = save_function
...
dispatch[types.MethodType] = save_instancemethod
...
dispatch[types.BuiltinFunctionType] = save_builtin_function
...
dispatch[classmethod] = save_classmethod
dispatch[staticmethod] = save_classmethod
```

I believe the name `save_global` is a holdover from the base `Pickler` class, which calls `self.save_global` in some base class methods that we call into via `super`, so renaming this isn't straightforward.

Reply:
I have pushed a new commit to simplify that function even further and all tests still pass. However, I am not sure we are not breaking edge cases in third party libraries and applications.

Reply:
I have run the test suite of both loky and joblib against this branch all tests pass as well so I am pretty confident that the changes are fine.

Reply:
Can you check whether `save_dynamic_class` really gets a function as input and, if so, change the name or docstring?

Reply:
@pitrou I'm pretty confident `save_dynamic_class` can never be called with a function as input.  It accesses `obj.__bases__` unconditionally, which doesn't exist on function objects.